pkg/standalone-utils/contracts/relayer/EulerWrapping.sol

```diff
@@ -0,0 +1,130 @@
+     * @param eulerProtocol - the address of the euler protocol
+     */
+    constructor(address eulerProtocol) {
+        _eulerProtocol = eulerProtocol;
+    }
+
+    function wrapEuler(
+        IEulerToken wrappedToken,
+        address sender,
+        address recipient,
+        uint256 amount,
+        uint256 outputReference
+    ) external payable {
+        if (_isChainedReference(amount)) {
+            amount = _getChainedReferenceValue(amount);
+        }
+
+        IERC20 underlying = IERC20(wrappedToken.underlyingAsset());
+
+        // The wrap caller is the implicit sender of tokens, so if the goal is for the tokens
+        // to be sourced from outside the relayer, we must first pull them here.
+        if (sender != address(this)) {
+            require(sender == msg.sender, "Incorrect sender");
+            _pullToken(sender, underlying, amount);
+        }
+
+        underlying.safeApprove(_eulerProtocol, amount);
+
+        // To calculate balances of the wrappedToken in the Relayer
+        IERC20 wrappedTokenErc20 = IERC20(address(wrappedToken));
+
+        // Deposit MainToken into EulerToken
+        // 0 for the Euler primary account
+        wrappedToken.deposit(0, amount);
+
+        uint256 withdrawnWrappedAmount = wrappedTokenErc20.balanceOf(address(this));
+
+        if (recipient != address(this)) {
+            wrappedTokenErc20.safeApprove(address(this), withdrawnWrappedAmount);
+            wrappedTokenErc20.safeTransferFrom(address(this), recipient, withdrawnWrappedAmount);
+        }
+
+        if (_isChainedReference(outputReference)) {
+            _setChainedReferenceValue(outputReference, withdrawnWrappedAmount);
+        }
+    }
+
+    function unwrapEuler(
+        IEulerToken wrappedToken,
+        address sender,
+        address recipient,
+        uint256 amount,
+        uint256 outputReference
+    ) external payable {
+        if (_isChainedReference(amount)) {
+            amount = _getChainedReferenceValue(amount);
+        }
+
+        // The unwrap caller is the implicit sender of tokens, so if the goal is for the tokens
+        // to be sourced from outside the relayer, we must first pull them here.
+        if (sender != address(this)) {
+            require(sender == msg.sender, "Incorrect sender");
+            _pullToken(sender, IERC20(address(wrappedToken)), amount);
+        }
+
+        IERC20 mainToken = IERC20(wrappedToken.underlyingAsset());
+
+        // Euler offers two ways to withdraw. Either calculate the MainTokenOut via
+        // 1. MainTokenOut = wrappedToken.convertBalanceToUnderlying(WrappedTokenAmount); or
+        // 2. withdraw as Much MainToken as one gets when depositing WrappedTokenAmount back
+        // When using option 1, the possibility exists to have dust of WrappedToken in the relayer
+        // Therefor option 2 is chosen
+        // 0 for the Euler primary account
+        wrappedToken.withdraw(0, 2**256 - 1); //MAX_UINT forces option 2
```

Comment:
Maybe it's better to calculate this in another variable:

`uint256 MAX_UINT = 0 - 1;`

This will produce the same result, since unsigned ints don't accept negative values. Calculating 2^256 may be more expensive.

Reply:
By the way, 2^256 needs 257 bits to be stored, so it's probably evaluated as 0 anyway

Reply:
done